File: src/obsidian_sync/sync_agent/conflict.py
import logging
from datetime import datetime
from pathlib import Path, PurePosixPath

from obsidian_sync.domain.hashing import sha256_file
from obsidian_sync.domain.paths import safe_vault_destination
from obsidian_sync.sync_agent.atomic import write_bytes_atomic, write_text_atomic

SERVER_DELETED_PLACEHOLDER = '(file is deleted on the server)'
LOCAL_DELETED_PLACEHOLDER = '(file was deleted locally)'
_SERVER_REVISION_PREFIX = '- Server revision:'

_LOGGER = logging.getLogger(__name__)


def conflict_filename(path: str, *, device_id: str, moment: datetime) -> str:
    stem = PurePosixPath(path).stem
    timestamp = moment.strftime('%Y%m%d-%H%M%S')
    return f'{stem}.conflict.{device_id}.{timestamp}.md'


def write_conflict_file(
    vault_root: Path,
    path: str,
    *,
    device_id: str,
    client_base_revision: int,
    server_revision: int,
    local_content: str,
    server_content: str,
) -> Path:
    """Write a conflict copy next to the file and return its path.

    If a conflict file for the same path, device, and server revision already
    exists, it is reused instead of writing another timestamped copy. This
    prevents conflict-file proliferation on a persistently diverged path.
    """
    moment = datetime.now().astimezone()
    parent = PurePosixPath(path).parent
    filename = conflict_filename(path, device_id=device_id, moment=moment)
    relative = (
        filename
        if parent in (PurePosixPath('.'), PurePosixPath(''))
        else (f'{parent.as_posix()}/{filename}')
    )
    destination = safe_vault_destination(vault_root, relative)
    existing = _find_existing_conflict(
        destination.parent,
        stem=PurePosixPath(path).stem,
        device_id=device_id,
        server_revision=server_revision,
    )
    if existing is not None:
        _LOGGER.info(
            'reusing conflict file %s for %s (server revision %s)',
            existing.name,
            path,
            server_revision,
        )
        return existing
    body = (
        '# Sync Conflict\n\n'
        f'- Path: {path}\n'
        f'- Device: {device_id}\n'
        f'- Client base revision: {client_base_revision}\n'
        f'- Server revision: {server_revision}\n'
        f'- Created at: {moment.isoformat()}\n\n'
        '## Local Version\n\n'
        f'{local_content}\n\n'
        '## Server Version\n\n'
        f'{server_content}\n'
    )
    write_text_atomic(destination, body)
    return destination


def local_backup_filename(path: str, *, device_id: str, moment: datetime) -> str:
    """Build the filename for a remote-wins local backup.

    The name deliberately contains the ``.conflict.`` marker used by
    ``domain.sync_rules.is_conflict_file`` so the backup is excluded from
    scanning, pushing, and server-side vectorizing just like a manual
    conflict file, while still being distinguishable from one (different
    prefix, so it never collides with manual-conflict dedup lookups).
    """
    stem = PurePosixPath(path).stem
    timestamp = moment.strftime('%Y%m%d-%H%M%S')
    return f'{stem}.local-backup.conflict.{device_id}.{timestamp}.md'


def write_local_backup_file(
    vault_root: Path,
    path: str,
    *,
    device_id: str,
    server_revision: int,
    local_content: str,
) -> Path:
    """Preserve local content discarded by a remote-wins auto-resolution.

    Called instead of overwriting or deleting local content silently: the
    server keeps a full version history so the server side is always
    recoverable, but the locally-discarded content only survives here.
    """
    moment = datetime.now().astimezone()
    parent = PurePosixPath(path).parent
    filename = local_backup_filename(path, device_id=device_id, moment=moment)
    relative = (
        filename
        if parent in (PurePosixPath('.'), PurePosixPath(''))
        else (f'{parent.as_posix()}/{filename}')
    )
    destination = safe_vault_destination(vault_root, relative)
    body = (
        '# Sync Conflict (auto-resolved: remote-wins)\n\n'
        f'- Path: {path}\n'
        f'- Device: {device_id}\n'
        f'- Server revision applied: {server_revision}\n'
        f'- Created at: {moment.isoformat()}\n\n'
        '## Local Version (discarded)\n\n'
        f'{local_content}\n'
    )
    write_text_atomic(destination, body)
    return destination


def _find_existing_conflict(
    directory: Path,
    *,
    stem: str,
    device_id: str,
    server_revision: int,
) -> Path | None:
    if not directory.exists():
        return None
    prefix = f'{stem}.conflict.{device_id}.'
    for candidate in sorted(directory.iterdir()):
        name = candidate.name
        if not (name.startswith(prefix) and name.endswith('.md')):
            continue
        if _parse_server_revision(candidate) == server_revision:
            return candidate
    return None


def _parse_server_revision(path: Path) -> int | None:
    try:
        text = path.read_text(encoding='utf-8')
    except OSError:
        return None
    for line in text.splitlines():
        if line.startswith(_SERVER_REVISION_PREFIX):
            value = line[len(_SERVER_REVISION_PREFIX) :].strip()
            try:
                return int(value)
            except ValueError:
                return None
    return None


def _conflict_relative_path(path: str, filename: str) -> str:
    parent = PurePosixPath(path).parent
    if parent in (PurePosixPath('.'), PurePosixPath('')):
        return filename
    return f'{parent.as_posix()}/{filename}'


def attachment_conflict_filename(path: str, *, device_id: str, moment: datetime) -> str:
    """Build a conflict filename for an attachment, extension preserved.

    Unlike ``conflict_filename`` (markdown), the original extension is kept
    instead of being replaced with ``.md`` (e.g. ``img.conflict.laptop.
    20260707-120000.png``), since the file must remain openable as its
    original type.
    """
    stem = PurePosixPath(path).stem
    extension = PurePosixPath(path).suffix
    timestamp = moment.strftime('%Y%m%d-%H%M%S')
    return f'{stem}.conflict.{device_id}.{timestamp}{extension}'


def write_attachment_conflict_file(
    vault_root: Path,
    path: str,
    *,
    device_id: str,
    content: bytes,
    content_hash: str,
) -> Path:
    """Write a side-by-side attachment conflict copy, extension preserved.

    Binary content cannot be textually merged or embedded in a single
    conflict report the way markdown conflicts embed both versions as
    readable sections, so the resolution here is a side-by-side copy: the
    original file at ``path`` is left untouched, and ``content`` (the
    "other" version relevant to the conflict -- typically the server's, or
    the local one when the server has nothing) is written to a sibling file
    whose name keeps the ``.conflict.`` marker (excluding it from future
    scanning/pushing/vectorizing) and the original extension.

    Deduplicated the same way markdown conflicts are: if a conflict file for
    the same (path, device) with matching content already exists, it is
    reused instead of writing another timestamped copy. Markdown dedup keys
    off an embedded ``- Server revision:`` text line; binary files have no
    such line, so this keys off a direct content-hash comparison instead.
    """
    moment = datetime.now().astimezone()
    filename = attachment_conflict_filename(path, device_id=device_id, moment=moment)
    destination = safe_vault_destination(
        vault_root, _conflict_relative_path(path, filename)
    )
    existing = _find_existing_attachment_conflict(
        destination.parent,
        stem=PurePosixPath(path).stem,
        extension=PurePosixPath(path).suffix,
        device_id=device_id,
        content_hash=content_hash,
    )
    if existing is not None:
        _LOGGER.info('reusing attachment conflict file %s for %s', existing.name, path)
        return existing
    write_bytes_atomic(destination, content)
    return destination


def attachment_local_backup_filename(
    path: str, *, device_id: str, moment: datetime
) -> str:
    stem = PurePosixPath(path).stem
    extension = PurePosixPath(path).suffix
    timestamp = moment.strftime('%Y%m%d-%H%M%S')
    return f'{stem}.local-backup.conflict.{device_id}.{timestamp}{extension}'


def write_attachment_local_backup_file(
    vault_root: Path,
    path: str,
    *,
    device_id: str,
    local_content: bytes,
) -> Path:
    """Preserve local attachment bytes discarded by a remote-wins resolution.

    Binary counterpart of ``write_local_backup_file``: no dedup (matching
    that function), since each remote-wins resolution is a distinct event
    worth preserving on its own.
    """
    moment = datetime.now().astimezone()
    filename = attachment_local_backup_filename(
        path, device_id=device_id, moment=moment
    )
    destination = safe_vault_destination(
        vault_root, _conflict_relative_path(path, filename)
    )
    write_bytes_atomic(destination, local_content)
    return destination


def _find_existing_attachment_conflict(
    directory: Path,
    *,
    stem: str,
    extension: str,
    device_id: str,
    content_hash: str,
) -> Path | None:
    if not directory.exists():
        return None
    prefix = f'{stem}.conflict.{device_id}.'
    for candidate in sorted(directory.iterdir()):
        name = candidate.name
        if not (name.startswith(prefix) and name.endswith(extension)):
            continue
        try:
            if sha256_file(candidate) == content_hash:
                return candidate
        except OSError:
            continue
    return None

File: internal/syncagent/conflict/conflict.go
package conflict

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"path"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/Qulip/obsidian-sync/internal/syncagent/vaultfs"
)

const (
	ServerDeletedPlaceholder = "(file is deleted on the server)"
	LocalDeletedPlaceholder  = "(file was deleted locally)"
)

const serverRevisionPrefix = "- Server revision:"

type Request struct {
	VaultRoot          string
	Path               string
	DeviceID           string
	ClientBaseRevision int
	ServerRevision     int
	LocalContent       string
	ServerContent      string
	Now                time.Time
}

func Filename(filePath string, deviceID string, moment time.Time) string {
	stem := strings.TrimSuffix(path.Base(filePath), path.Ext(filePath))
	timestamp := moment.Format("20060102-150405")
	return fmt.Sprintf("%s.conflict.%s.%s.md", stem, deviceID, timestamp)
}

// AttachmentFilename mirrors Filename but keeps the original extension:
// binary content can't show both versions inline like a markdown conflict
// doc does, so the "other" version is written out as its own file (see
// docs/sync-agent.md, "충돌 파일 (확장자 유지)").
func AttachmentFilename(filePath string, deviceID string, moment time.Time) string {
	stem := strings.TrimSuffix(path.Base(filePath), path.Ext(filePath))
	timestamp := moment.Format("20060102-150405")
	return fmt.Sprintf("%s.conflict.%s.%s%s", stem, deviceID, timestamp, path.Ext(filePath))
}

type BinaryRequest struct {
	VaultRoot string
	Path      string
	DeviceID  string
	// Content is the "other" version's raw bytes: the server's bytes on a
	// write conflict, or the local bytes on a server-deleted conflict.
	Content []byte
	Now     time.Time
}

// WriteBinaryFile writes an attachment conflict copy, deduplicated by
// content hash since binary files have no text marker to parse the way
// WriteFile's "- Server revision:" line does for markdown.
func WriteBinaryFile(req BinaryRequest) (string, error) {
	moment := req.Now
	if moment.IsZero() {
		moment = time.Now().Local()
	}

	destination, err := resolveConflictDestination(req.VaultRoot, req.Path, AttachmentFilename(req.Path, req.DeviceID, moment))
	if err != nil {
		return "", err
	}

	contentHash := hashBytes(req.Content)
	existing, err := findExistingBinary(existingBinaryConflictQuery{
		directory:   filepath.Dir(destination),
		stem:        strings.TrimSuffix(path.Base(req.Path), path.Ext(req.Path)),
		ext:         path.Ext(req.Path),
		deviceID:    req.DeviceID,
		contentHash: contentHash,
	})
	if err != nil {
		return "", err
	}
	if existing != "" {
		return existing, nil
	}

	if err := os.MkdirAll(filepath.Dir(destination), 0o755); err != nil {
		return "", fmt.Errorf("create conflict directory: %w", err)
	}
	if err := writeAtomic(destination, req.Content); err != nil {
		return "", err
	}
	return destination, nil
}

type existingBinaryConflictQuery struct {
	directory   string
	stem        string
	ext         string
	deviceID    string
	contentHash string
}

func findExistingBinary(query existingBinaryConflictQuery) (string, error) {
	entries, err := os.ReadDir(query.directory)
	if os.IsNotExist(err) {
		return "", nil
	}
	if err != nil {
		return "", fmt.Errorf("read conflict directory: %w", err)
	}
	prefix := fmt.Sprintf("%s.conflict.%s.", query.stem, query.deviceID)
	for _, entry := range entries {
		if entry.IsDir() || entry.Type()&os.ModeSymlink != 0 {
			continue
		}
		name := entry.Name()
		if !strings.HasPrefix(name, prefix) || !strings.HasSuffix(name, query.ext) {
			continue
		}
		candidate := filepath.Join(query.directory, name)
		data, err := os.ReadFile(candidate)
		if err != nil {
			continue
		}
		if hashBytes(data) == query.contentHash {
			return candidate, nil
		}
	}
	return "", nil
}

func hashBytes(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

func WriteFile(req Request) (string, error) {
	moment := req.Now
	if moment.IsZero() {
		moment = time.Now().Local()
	}

	destination, err := destinationPath(req, moment)
	if err != nil {
		return "", err
	}

	existing, err := findExisting(existingConflictQuery{
		directory:      filepath.Dir(destination),
		stem:           strings.TrimSuffix(path.Base(req.Path), path.Ext(req.Path)),
		deviceID:       req.DeviceID,
		serverRevision: req.ServerRevision,
	})
	if err != nil {
		return "", err
	}
	if existing != "" {
		return existing, nil
	}

	if err := os.MkdirAll(filepath.Dir(destination), 0o755); err != nil {
		return "", fmt.Errorf("create conflict directory: %w", err)
	}
	body := formatBody(req, moment)
	if err := writeTextAtomic(destination, body); err != nil {
		return "", err
	}
	return destination, nil
}

func destinationPath(req Request, moment time.Time) (string, error) {
	return resolveConflictDestination(req.VaultRoot, req.Path, Filename(req.Path, req.DeviceID, moment))
}

func resolveConflictDestination(vaultRoot string, filePath string, name string) (string, error) {
	parent := path.Dir(filePath)
	relative := name
	if parent != "." && parent != "/" {
		relative = path.Join(parent, name)
	}
	cleanRelative := filepath.Clean(filepath.FromSlash(relative))
	if filepath.IsAbs(cleanRelative) || cleanRelative == ".." || strings.HasPrefix(cleanRelative, ".."+string(filepath.Separator)) {
		return "", fmt.Errorf("conflict path escapes vault: %s", filePath)
	}
	destination, err := vaultfs.SafePath(vaultRoot, relative)
	if err != nil {
		return "", fmt.Errorf("resolve conflict path: %w", err)
	}
	return destination, nil
}

type existingConflictQuery struct {
	directory      string
	stem           string
	deviceID       string
	serverRevision int
}

func findExisting(query existingConflictQuery) (string, error) {
	entries, err := os.ReadDir(query.directory)
	if os.IsNotExist(err) {
		return "", nil
	}
	if err != nil {
		return "", fmt.Errorf("read conflict directory: %w", err)
	}
	prefix := fmt.Sprintf("%s.conflict.%s.", query.stem, query.deviceID)
	for _, entry := range entries {
		if entry.IsDir() || entry.Type()&os.ModeSymlink != 0 {
			continue
		}
		name := entry.Name()
		if !strings.HasPrefix(name, prefix) || !strings.HasSuffix(name, ".md") {
			continue
		}
		candidate := filepath.Join(query.directory, name)
		revision, ok := parseServerRevision(candidate)
		if ok && revision == query.serverRevision {
			return candidate, nil
		}
	}
	return "", nil
}

func parseServerRevision(filePath string) (int, bool) {
	body, err := os.ReadFile(filePath)
	if err != nil {
		return 0, false
	}
	for _, line := range strings.Split(string(body), "\n") {
		if !strings.HasPrefix(line, serverRevisionPrefix) {
			continue
		}
		value := strings.TrimSpace(strings.TrimPrefix(line, serverRevisionPrefix))
		revision, err := strconv.Atoi(value)
		if err != nil {
			return 0, false
		}
		return revision, true
	}
	return 0, false
}

func formatBody(req Request, moment time.Time) string {
	return fmt.Sprintf(
		"# Sync Conflict\n\n"+
			"- Path: %s\n"+
			"- Device: %s\n"+
			"- Client base revision: %d\n"+
			"- Server revision: %d\n"+
			"- Created at: %s\n\n"+
			"## Local Version\n\n"+
			"%s\n\n"+
			"## Server Version\n\n"+
			"%s\n",
		req.Path,
		req.DeviceID,
		req.ClientBaseRevision,
		req.ServerRevision,
		moment.Format(time.RFC3339),
		req.LocalContent,
		req.ServerContent,
	)
}

func writeTextAtomic(destination string, content string) error {
	return writeAtomic(destination, []byte(content))
}

func writeAtomic(destination string, data []byte) (err error) {
	temp, err := os.CreateTemp(filepath.Dir(destination), "."+filepath.Base(destination)+".tmp-*")
	if err != nil {
		return fmt.Errorf("create conflict temp file: %w", err)
	}
	tempName := temp.Name()
	defer func() {
		if err != nil {
			_ = os.Remove(tempName)
		}
	}()
	if _, err := temp.Write(data); err != nil {
		_ = temp.Close()
		return fmt.Errorf("write conflict temp file: %w", err)
	}
	if err := temp.Close(); err != nil {
		return fmt.Errorf("close conflict temp file: %w", err)
	}
	if err := os.Rename(tempName, destination); err != nil {
		return fmt.Errorf("replace conflict file: %w", err)
	}
	return nil
}
